src/Nethermind/Nethermind.Network/NetworkStorage.cs

```diff
@@ -16,52 +20,93 @@ namespace Nethermind.Network
 {
     public class NetworkStorage : INetworkStorage
     {
+        private readonly object _lock = new();
         private readonly IFullDb _fullDb;
         private readonly ILogger _logger;
+        private readonly List<NetworkNode> _nodesList = new();
+        private readonly HashSet<PublicKey> _nodePublicKeys = new();
         private long _updateCounter;
         private long _removeCounter;
+        private NetworkNode[] _nodes;
 
         public NetworkStorage(IFullDb? fullDb, ILogManager? logManager)
         {
             _fullDb = fullDb ?? throw new ArgumentNullException(nameof(fullDb));
             _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
         }
 
-        public int PersistedNodesCount
+        public int PersistedNodesCount => GetPersistedNodes().Length;
+
+        public NetworkNode[] GetPersistedNodes()
         {
-            get
-            {
-                return _fullDb.Values.Count;
-            }
+            NetworkNode[] nodes = _nodes;
+            return nodes is not null ? nodes : GenerateNodes();
         }
 
-        public NetworkNode[] GetPersistedNodes()
+        private NetworkNode[] GenerateNodes()
         {
-            List<NetworkNode> nodes = new();
-            foreach (byte[]? nodeRlp in _fullDb.Values)
+            NetworkNode[] nodes = Volatile.Read(ref _nodes);
+            lock (_lock)
             {
-                if (nodeRlp is null)
+                if (_nodes != nodes)
                 {
-                    continue;
+                    // Already updated
+                    return _nodes;
                 }
 
-                try
+                _nodePublicKeys.Clear();
+                _nodesList.Clear();
+                foreach (byte[]? nodeRlp in _fullDb.Values)
                 {
-                    nodes.Add(GetNode(nodeRlp));
-                }
-                catch (Exception e)
-                {
-                    if (_logger.IsDebug) _logger.Debug($"Failed to add one of the persisted nodes (with RLP {nodeRlp.ToHexString()}), {e.Message}");
+                    if (nodeRlp is null)
+                    {
+                        continue;
+                    }
+
+                    try
+                    {
+                        NetworkNode node = GetNode(nodeRlp);
+                        _nodesList.Add(node);
+                        _nodePublicKeys.Add(node.NodeId);
```

Comment:
We could potentially reuse already existing nodes instead of creating new ones all the time

Reply:
`RemoveNode` and `UpdateNode` don't set the `_nodes` array to `null` but just replace it so `GenerateNodes()` should only be creating the nodes from scratch once during app lifetime (i.e. this bit of code) and `UpdateNode` gets passed in the preconstructed `NetworkNode` so it would require a fair bit of refactoring to reuse the nodes from `RemoveNode` 🤔

Reply:
So rlp starts with public key, which could be decoded on stack, then it could be reused from existing node instead of creating new one.

We can say it is out of scope of this PR and try to do it in some next one.

Reply:
Yeah I think would be good to look at that; as well as wider, as are a lot of temporary `byte[]` created for `Rlp` only to be fed into `KeccakHash.Compute` then thrown away; but the temp `byte[]` could be bypassed by feeding direct into `KeccakHash` which would also reduce copies by one